src/cliffordsynthesis/GateEncoding.cpp

```diff
@@ -0,0 +1,364 @@
+/*
+* This file is part of the MQT QMAP library which is released under the MIT license.
+* See file README.md or go to https://www.cda.cit.tum.de/research/ibm_qx_mapping/ for more information.
+*/
+
+#include "cliffordsynthesis/GateEncoding.hpp"
+void cs::GateEncoding::makeSingleGateEncoding(const SynthesisData& data) {
+    auto changes = logicbase::LogicTerm(true);
+    // CONSISTENCY
+    // One gate per qubit, per step
+    for (std::size_t gateStep = 1U; gateStep < data.timesteps + 1U; ++gateStep) {
+        std::vector<logicbase::LogicTerm> vars{};
+        auto                              aIt = data.qubitChoice.cbegin();
+        for (std::size_t a = 0U; a < data.nqubits; ++a) {
+            for (auto gate: Gates::SINGLE_QUBIT) {
+                vars.emplace_back(data.gS[gateStep][Gates::toIndex(gate)][a]);
+            }
+            auto bIt = data.qubitChoice.cbegin();
+            for (std::size_t b = 0; b < data.nqubits; ++b) {
+                const auto q0 = *aIt;
+                const auto q1 = *bIt;
+                ++bIt;
+                if (a == b || data.reducedCM.find({q0, q1}) == data.reducedCM.end()) {
+                    continue;
+                }
+                vars.emplace_back(data.gTwoQubit[gateStep][a][b]);
+            }
+            ++aIt;
+        }
+        data.lb->assertFormula(encodings::exactlyOneCmdr(
+                encodings::groupVars(vars, static_cast<std::size_t>(vars.size() / 2U)),
+                logicbase::LogicTerm::noneTerm(), data.lb.get()));
+    }
+
+    // GATE CONSTRAINTS
+    for (unsigned int gateStep = 1; gateStep < data.timesteps + 1; ++gateStep) {
+        auto aIt = data.qubitChoice.cbegin();
+        for (unsigned int a = 0; a < data.nqubits; ++a) {
+            // NO GATE
+            changes = (data.x[gateStep][a] == data.x[gateStep - 1][a]);
+            changes = changes && (data.z[gateStep][a] == data.z[gateStep - 1][a]);
+
+            for (unsigned int b = 0; b < data.nqubits; ++b) {
+                if (a == b) {
+                    continue;
+                }
+                changes = changes && (data.x[gateStep][b] == data.x[gateStep - 1][b]);
+                changes = changes && (data.z[gateStep][b] == data.z[gateStep - 1][b]);
+            }
+
+            changes = changes && (data.r[gateStep] == data.r[gateStep - 1]);
+            changes = logicbase::LogicTerm::implies(data.gS[gateStep][0][a], changes);
+            data.lb->assertFormula(changes);
```

Comment:
please split this huge function into multiple smaller ones that each have a dedicated purpose.

Also there seems to be lots of code duplication between these individual functions. Try to reduce this as much as possible by encapsulating shared functionality.

As of now, this file is hardly readable and lacks structure. this makes it really hard to check all these constraints for correctness.